### PR TITLE
j2735_convertor and cpp_message updates for BSM Part II Content and Regional Extensions

### PR DESCRIPTION
**Pull request by JonSmet**
# PR Details
## Description

This PR includes the following updates:
- j2735_convertor updates to convert BSM Part II Content
- j2735_convertor updates to convert BSM Regional Extensions
- cpp_message updates to encode and decode BSM Part II Content
- cpp_message updates to encode and decode BSM Regional Extensions

While the lines of code included in this PR seems substantial, the changes are mostly straightforward and repetitive. Overall, the updates to support BSM Part II Content and BSM Regional Extensions includes over 130 nested messages.

Additionally, during this PR it was found that the vlm/asn1c [(Link Here)](https://github.com/mouse07410/asn1c/tree/vlm_master/asn1c) repository does not support compilation of BSM Part II Content or Regional Extensions. In an effort to enable this, the mouse07410/asn1c forked repository [(Link Here)](https://github.com/mouse07410/asn1c/tree/vlm_master/asn1c) has been used instead, since it includes the necessary fixes for these components. The instructions for generating .c, .h. and .so files from a .asn file for CARMA Messenger are located within CARMA Messenger under lib_asn1c/third_party_lib/libasn1c_update.txt, and have been updated to use this new forked repository.

During this PR, some missing max/min fields and some incorrect message fields were found within carma-msgs packages. These have been fixed in the related carma-msgs PR #199: [Click Here](https://github.com/usdot-fhwa-stol/carma-msgs/pull/199)

## How Has This Been Tested?

All unit tests are successful.

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by saina-ramyar** [comment]
This is only j2735 <--> cav_msgs? Not updating the cpp msg node? Just clarifying since I see a mention of encode/decode

**Comment by JonSmet** [comment]
> This is only j2735 <--> cav_msgs? Not updating the cpp msg node? Just clarifying since I see a mention of encode/decode

@saina-ramyar Currently only the j2735_convertor updates are included in this draft PR, but later today I will be adding commits to cpp_message for encoding/decoding.

File: carma-messenger-core/j2735_convertor/src/bsm_convertor.cpp [hunk 1]
```diff
@@ -22,6 +22,162 @@
...
+        out_msg.initial_position.presence_vector |= in_msg.initial_position.presence_vector;
+        out_msg.initial_position.lat.latitude = in_msg.initial_position.lat.latitude / units::TENTH_MICRO_DEG_PER_DEG;
+        out_msg.initial_position.lon.longitude = in_msg.initial_position.lon.longitude / units::TENTH_MICRO_DEG_PER_DEG;
+        out_msg.initial_position.elevation.elevation = in_msg.initial_position.elevation.elevation / units::DECI_M_PER_M;
+        out_msg.initial_position.utc_time = in_msg.initial_position.utc_time;
+        out_msg.initial_position.time_confidence = in_msg.initial_position.time_confidence;
+        out_msg.initial_position.pos_confidence = in_msg.initial_position.pos_confidence;
+        out_msg.initial_position.speed_confidence = in_msg.initial_position.speed_confidence;
+
+        out_msg.initial_position.lat.unavailable = 0;
```
[the remark this answers is not in the document]
Comment (JonSmet) [reply carma-messenger-core/j2735_convertor/src/bsm_convertor.cpp]: Good catch, fixed.

File: carma-messenger-core/j2735_convertor/src/bsm_convertor.cpp [hunk 2]
```diff
@@ -128,13 +815,703 @@ void BSMConvertor::convert(const j2735_v2x_msgs::msg::BSMCoreData& in_msg, carma
...
+        }
+
+        out_msg.push_back(out_position_3d);
+    }
+}
+
+void BSMConvertor::convert(const carma_v2x_msgs::msg::PathPrediction& in_msg, j2735_v2x_msgs::msg::PathPrediction& out_msg)
+{
+  out_msg.radius_of_curvature = in_msg.radius_of_curvature * units::CENTI_DEG_PER_DEG;
+  out_msg.confidence = in_msg.confidence * 200;
```
[the remark this answers is not in the document]
Comment (JonSmet) [reply carma-messenger-core/j2735_convertor/src/bsm_convertor.cpp]: Fixed.

File: carma-messenger-core/j2735_convertor/src/bsm_convertor.cpp [hunk 3]
```diff
@@ -128,13 +815,703 @@ void BSMConvertor::convert(const j2735_v2x_msgs::msg::BSMCoreData& in_msg, carma
...
+            out_msg.vehicle_data.bumpers.rear.bumper_height = in_msg.vehicle_data.bumpers.rear.bumper_height * units::CM_PER_M;
+        }
+
+        // vehicle_data.mass
+        if(in_msg.vehicle_data.presence_vector & carma_v2x_msgs::msg::VehicleData::HAS_MASS){
+            if(in_msg.vehicle_data.mass.unavailable){
+                out_msg.vehicle_data.mass.vehicle_mass = j2735_v2x_msgs::msg::VehicleMass::MASS_UNAVAILABLE;
+            }
+            else{
+                // NOTE: Refer to J2735 ASN Specification for more details regarding this conversion.
```
[the remark this answers is not in the document]
Comment (JonSmet) [reply carma-messenger-core/j2735_convertor/src/bsm_convertor.cpp]: Good idea, added this in the comments.

File: carma-messenger-core/cpp_message/src/BSM_Message.cpp [hunk 4]
```diff
@@ -89,36 +434,805 @@ namespace cpp_message
...
+                    output.part_ii.push_back(part_ii_output);
+                }
+            }
+
+            // Decode regional list
+            if (message->value.choice.BasicSafetyMessage.regional){
+                // TODO
+            }
             
+            ASN_STRUCT_FREE(asn_DEF_MessageFrame, message);
```
[the remark this answers is not in the document]
Comment (JonSmet) [reply carma-messenger-core/cpp_message/src/BSM_Message.cpp:1275]: Yes, added this as a comment.

File: carma-messenger-core/cpp_message/test/test_BSM.cpp [hunk 5]
```diff
@@ -450,6 +468,13 @@ TEST(BSMTest, testEncodeDecodeBSM)
     }
     j2735_v2x_msgs::msg::BSM result = res_decoded.get();
 
+    // if(result.regional.size() == message.regional.size()){
```
Comment (saina-ramyar) [inline carma-messenger-core/cpp_message/test/test_BSM.cpp:471]: remove the commented code if not needed

**Review by saina-ramyar — APPROVED** [review]

File: carma-messenger-core/cpp_message/test/test_BSM.cpp (hunk 5 above)
Comment (JonSmet) [reply carma-messenger-core/cpp_message/test/test_BSM.cpp]: Fixed.

**Review by saina-ramyar — APPROVED** [review]

**Pull request merged**